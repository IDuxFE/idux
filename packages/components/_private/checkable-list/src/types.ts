/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { VirtualScrollToFn } from '@idux/cdk/scroll'
import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray, VKey } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes, PropType } from 'vue'

export interface CheckableListData {
  key?: VKey
  label?: string
  disabled?: boolean
  [key: string]: unknown
}

export interface CheckableListScroll {
  height?: string | number
  fullHeight?: boolean
}

export const checkableListProps = {
  dataSource: Array as PropType<CheckableListData[]>,
  checkable: {
    type: Boolean,
    default: true,
  },
  removable: {
    type: Boolean,
    default: false,
  },
  checked: Function as PropType<(item: CheckableListData) => boolean>,
  customAdditional: { type: Function as PropType<CheckableListCustomAdditional>, default: undefined },
  disabled: Function as PropType<(item: CheckableListData) => boolean>,
  getKey: Function as PropType<(item: CheckableListData) => VKey>,
  labelKey: String,
  virtual: {
    type: Boolean,
    default: false,
  },
  scroll: Object as PropType<CheckableListScroll>,
  onCheckChange: [Function, Array] as PropType<MaybeArray<(item: CheckableListData, checked: boolean) => void>>,
  onRemove: [Function, Array] as PropType<MaybeArray<(item: CheckableListData) => void>>,
  onScroll: [Function, Array] as PropType<MaybeArray<(evt: Event) => void>>,
  onScrolledChange: [Function, Array] as PropType<
    MaybeArray<(startIndex: number, endIndex: number, visibleData: unknown[]) => void>
  >,
  onScrolledBottom: [Function, Array] as PropType<MaybeArray<() => void>>,
} as const

export const checkableListItemProps = {
  checked: {
    type: Boolean,
    default: false,
  },
  checkable: {
    type: Boolean,
    default: true,
  },
  removable: {
    type: Boolean,
    default: false,
  },
  disabled: {
    type: Boolean,
    default: false,
  },
  label: String,
  value: {
    type: [String, Number, Symbol] as PropType<VKey>,
    required: true,
  },
  onCheckChange: [Function, Array] as PropType<MaybeArray<(checked: boolean) => void>>,
  onRemove: [Function, Array] as PropType<MaybeArray<() => void>>,
} as const

export interface CheckableListApi {
  scrollTo: VirtualScrollToFn
}

export type CheckableListProps = ExtractInnerPropTypes<typeof checkableListProps>
export type CheckableListItemProps = ExtractInnerPropTypes<typeof checkableListItemProps>
export type CheckableListPublicProps = ExtractPublicPropTypes<typeof checkableListProps>
export type CheckableListComponent = DefineComponent<
  Omit<HTMLAttributes, keyof CheckableListPublicProps> & CheckableListPublicProps,
  CheckableListApi
>
export type CheckableListInstance = InstanceType<DefineComponent<CheckableListProps, CheckableListApi>>

export type CheckableListCustomAdditional = (options: {
  data: CheckableListData
  index: number
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
}) => Record<string, any> | undefined
