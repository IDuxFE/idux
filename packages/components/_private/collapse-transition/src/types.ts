/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes } from 'vue'

import { IxPropTypes } from '@idux/cdk/utils'

export type CollapseTransitionMode = 'height' | 'width'

export const collapseTransitionProps = {
  appear: IxPropTypes.bool.def(false),
  name: IxPropTypes.string,
  mode: IxPropTypes.oneOf<CollapseTransitionMode>(['height', 'width']).def('height'),
  onAfterEnter: IxPropTypes.emit(),
  onAfterLeave: IxPropTypes.emit(),
}

export type CollapseTransitionProps = ExtractInnerPropTypes<typeof collapseTransitionProps>
export type CollapseTransitionPublicProps = ExtractPublicPropTypes<typeof collapseTransitionProps>
export type CollapseTransitionComponent = DefineComponent<
  Omit<HTMLAttributes, keyof CollapseTransitionPublicProps> & CollapseTransitionPublicProps
>
export type CollapseTransitionInstance = InstanceType<DefineComponent<CollapseTransitionProps>>
