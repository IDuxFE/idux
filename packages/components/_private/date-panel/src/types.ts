/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes, PropType } from 'vue'

export const datePanelProps = {
  cellTooltip: Function as PropType<(cell: { value: Date; disabled: boolean }) => string | void>,
  disabledDate: Function as PropType<(date: Date) => boolean>,
  type: {
    type: String as PropType<DatePanelType>,
    default: 'date',
  },
  value: [Date, Array] as PropType<Date | (Date | undefined)[]>,
  activeDate: Date,
  visible: {
    type: Boolean,
    default: undefined,
  },
  isSelecting: {
    type: Boolean,
    default: undefined,
  },

  onCellClick: [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
  onCellMouseenter: [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
  'onUpdate:activeDate': [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
}

export type DatePanelProps = ExtractInnerPropTypes<typeof datePanelProps>
export type DatePanelPublicProps = ExtractPublicPropTypes<typeof datePanelProps>
export type DatePanelComponent = DefineComponent<
  Omit<HTMLAttributes, keyof DatePanelPublicProps> & DatePanelPublicProps
>
export type DatePanelInstance = InstanceType<DefineComponent<DatePanelProps>>

export type DatePanelType = 'date' | 'week' | 'month' | 'quarter' | 'year'

// private
export const panelRowProps = {
  rowIndex: {
    type: Number,
    required: true,
  },
}

export const panelCellProps = {
  rowIndex: {
    type: Number,
    required: true,
  },
  cellIndex: {
    type: Number,
    required: true,
  },
}
