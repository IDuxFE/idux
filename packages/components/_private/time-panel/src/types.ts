/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes, PropType } from 'vue'

import { MaybeArray } from '@idux/cdk/utils'

export type AmPm = 'am' | 'pm'

export const baseTimePanelProps = {
  disabledHours: {
    type: Function as PropType<(selectedAmPm: AmPm | undefined) => number[]>,
    default: () => [],
  },
  disabledMinutes: {
    type: Function as PropType<(selectedHour: number | undefined, selectedAmPm: AmPm | undefined) => number[]>,
    default: () => [],
  },
  disabledSeconds: {
    type: Function as PropType<
      (selectedHour: number | undefined, selectedMinute: number | undefined, selectedAmPm: AmPm | undefined) => number[]
    >,
    default: () => [],
  },

  hideDisabledOptions: {
    type: Boolean,
    default: false,
  },
  hourStep: {
    type: Number,
    default: 1,
  },
  minuteStep: {
    type: Number,
    default: 1,
  },
  secondStep: {
    type: Number,
    default: 1,
  },
} as const

export const timePanelProps = {
  ...baseTimePanelProps,

  value: Date,
  activeValue: Date,
  visible: {
    type: Boolean,
    default: false,
  },
  hourEnabled: {
    type: Boolean,
    default: true,
  },
  minuteEnabled: {
    type: Boolean,
    default: true,
  },
  secondEnabled: {
    type: Boolean,
    default: true,
  },
  use12Hours: {
    type: Boolean,
    default: false,
  },

  // events
  'onUpdate:value': [Function, Array] as PropType<MaybeArray<(value: Date) => void>>,
  'onUpdate:activeValue': [Function, Array] as PropType<MaybeArray<(value: Date) => void>>,
  onChange: [Function, Array] as PropType<MaybeArray<(value: Date) => void>>,
} as const

export type TimePanelProps = ExtractInnerPropTypes<typeof timePanelProps>
export type TimePanelPublicProps = ExtractPublicPropTypes<typeof timePanelProps>
export type TimePanelComponent = DefineComponent<
  Omit<HTMLAttributes, keyof TimePanelPublicProps> & TimePanelPublicProps
>
export type TimePanelInstance = InstanceType<DefineComponent<TimePanelProps>>

// private
export interface TimePanelCell {
  value: number | 'am' | 'pm'
  disabled: boolean
}

export const timePanelColumnProps = {
  selectedValue: [String, Number] as PropType<string | number>,
  activeValue: {
    type: [String, Number] as PropType<string | number>,
    required: true,
  },
  options: {
    type: Array as PropType<TimePanelCell[]>,
    require: true,
  },
  visible: {
    type: Boolean,
    default: false,
  },

  // events
  onActiveChange: [Function, Array] as PropType<MaybeArray<(cell: TimePanelCell) => void>>,
} as const

export const timePanelCellProps = {
  disabled: {
    type: Boolean,
    default: false,
  },
  selected: {
    type: Boolean,
    default: false,
  },
  value: {
    type: [String, Number] as PropType<string | number>,
    required: true,
  },

  // events
  onActive: [Function, Array] as PropType<MaybeArray<(cell: TimePanelCell) => void>>,
} as const

export type BaseTimePanelProps = ExtractInnerPropTypes<typeof baseTimePanelProps>
export type TimePanelColumnProps = ExtractInnerPropTypes<typeof timePanelColumnProps>
export type TimePanelCellProps = ExtractInnerPropTypes<typeof timePanelCellProps>

export type TimePanelColumnType = 'hour' | 'minute' | 'second' | 'AM/PM'
