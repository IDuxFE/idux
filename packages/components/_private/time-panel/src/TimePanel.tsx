/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { TimePanelColumnProps } from './types'

import { computed, defineComponent, provide } from 'vue'

import { useDateConfig, useGlobalConfig } from '@idux/components/config'

import PanelColumn from './TimePanelColumn'
import { useOptions } from './composables/useOptions'
import { timePanelContext } from './tokens'
import { timePanelProps } from './types'

export default defineComponent({
  name: 'IxTimePanel',
  props: timePanelProps,
  setup(props) {
    const common = useGlobalConfig('common')
    const mergedPrefixCls = computed(() => `${common.prefixCls}-time-panel`)
    const dateConfig = useDateConfig()
    const { hourOptionsProps, minuteOptionsProps, secondOptionsProps, amPmOptionsProps } = useOptions(props, dateConfig)

    provide(timePanelContext, { mergedPrefixCls })

    const columns = computed(() => {
      const result: TimePanelColumnProps[] = []
      props.hourEnabled && result.push(hourOptionsProps.value)
      props.minuteEnabled && result.push(minuteOptionsProps.value)
      props.secondEnabled && result.push(secondOptionsProps.value)
      props.use12Hours && result.push(amPmOptionsProps.value)
      return result
    })

    return () => (
      <div class={`${mergedPrefixCls.value}`}>
        {columns.value.map((item, index) => (
          <PanelColumn key={index} {...item} visible={props.visible} />
        ))}
      </div>
    )
  },
})
