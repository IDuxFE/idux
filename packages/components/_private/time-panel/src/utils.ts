/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { TimePanelColumnType } from './types'
import type { DateConfig } from '@idux/components/config'

export function normalizeAmPm(hour: number, is12Hours = false): 'am' | 'pm' | undefined {
  if (!is12Hours) {
    return
  }

  return hour >= 12 ? 'pm' : 'am'
}

export function calculateViewHour(hour: number, is12Hours: boolean): number {
  if (is12Hours) {
    hour > 12 && (hour -= 12)

    // if is12Hours, there is no 0, it should be 12 am or 12 pm
    hour === 0 && (hour = 12)
  }
  return hour
}

export function getHourValue(hour: number, ampm?: 'am' | 'pm' | ''): number {
  if (ampm) {
    ampm = ampm.toLowerCase() as 'am' | 'pm' | ''
    if (ampm === 'am') {
      hour >= 12 && (hour -= 12)
    }
    if (ampm === 'pm') {
      hour < 12 && (hour += 12)
    }
  }

  return hour
}

export function calculateValue(
  dateConfig: DateConfig,
  dateNow: Date,
  type: TimePanelColumnType,
  is12Hours: boolean,
  value: number | 'am' | 'pm' | '',
): Date {
  const { get, set } = dateConfig
  const selectNumber = Number(value)
  const newDate = new Date(dateNow)

  switch (type) {
    case 'hour':
      return set(newDate, getHourValue(selectNumber, normalizeAmPm(get(dateNow, 'hour'), is12Hours)), 'hour')
    case 'minute':
      return set(newDate, selectNumber, 'minute')
    case 'second':
      return set(newDate, selectNumber, 'second')
    case 'AM/PM':
      return set(newDate, getHourValue(get(newDate, 'hour'), value.toString() as 'am' | 'pm' | ''), 'hour')
  }
}
