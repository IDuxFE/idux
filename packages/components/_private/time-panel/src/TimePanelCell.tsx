/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { computed, defineComponent, inject } from 'vue'

import { callEmit, isNumeric } from '@idux/cdk/utils'

import { timePanelContext } from './tokens'
import { timePanelCellProps } from './types'

export default defineComponent({
  props: timePanelCellProps,
  setup(props) {
    const { mergedPrefixCls } = inject(timePanelContext)!
    const classes = computed(() => {
      const prefixCls = `${mergedPrefixCls.value}-cell`
      return {
        [prefixCls]: true,
        [`${prefixCls}-disabled`]: props.disabled,
        [`${prefixCls}-selected`]: props.selected,
      }
    })

    const onClick = () => {
      if (!props.disabled && !props.selected) {
        callEmit(props.onChange, props.value)
      }
    }

    const displayValue = computed(() => displayFormat(props.value))

    return () => (
      <li class={classes.value} onClick={onClick}>
        {displayValue.value}
      </li>
    )
  },
})

function displayFormat(val: string | number) {
  return isNumeric(val) ? val.toString().padStart(2, '0') : val
}
