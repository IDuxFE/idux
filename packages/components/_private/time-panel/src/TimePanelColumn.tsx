/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { defineComponent, inject, nextTick, ref, watch, watchEffect } from 'vue'

import { callEmit } from '@idux/cdk/utils'

import PanelCell from './TimePanelCell'
import { useColumnEvents } from './composables/useColumnEvents'
import { usePanelScroll } from './composables/usePanelScroll'
import { timePanelContext } from './tokens'
import { type TimePanelCell, type TimePanelColumnProps, timePanelColumnProps } from './types'

export default defineComponent({
  props: timePanelColumnProps,
  setup(props: TimePanelColumnProps) {
    const { mergedPrefixCls } = inject(timePanelContext)!
    const listRef = ref<HTMLElement | undefined>(undefined)
    const { scrollToActive, getTargetByScrollTop, frameRunning } = usePanelScroll(props, listRef, mergedPrefixCls)

    function onActive(cell: TimePanelCell) {
      callEmit(props.onActiveChange, cell)
    }
    function handleScrollAdjust() {
      scrollToActive()
    }

    const {
      handleWheel,
      handleClick,
      handleScroll: _handleScroll,
    } = useColumnEvents(props, listRef, getTargetByScrollTop, frameRunning)
    const handleListWheel = (evt: WheelEvent) => {
      evt.preventDefault()
    }

    let isScrolling = false
    const handleScroll = () => {
      isScrolling = true
      _handleScroll()
      nextTick(() => {
        isScrolling = false
      })
    }

    watchEffect(() => {
      if (props.visible) {
        nextTick(() => scrollToActive(0))
      }
    })
    watch(
      () => props.activeValue,
      () => {
        !isScrolling && nextTick(scrollToActive)
      },
    )

    return () => (
      <div
        class={`${mergedPrefixCls.value}-column`}
        onMouseleave={handleScrollAdjust}
        onMouseenter={handleScrollAdjust}
        onWheel={handleWheel}
        onClick={handleClick}
      >
        <ul
          ref={listRef}
          class={`${mergedPrefixCls.value}-column-inner`}
          onScroll={handleScroll}
          onWheel={handleListWheel}
        >
          {props.options!.map((item, index) => {
            const { disabled, value } = item
            return (
              <PanelCell
                key={index}
                disabled={disabled}
                selected={props.selectedValue === value}
                value={value}
                onActive={onActive}
              />
            )
          })}
        </ul>
      </div>
    )
  },
})
