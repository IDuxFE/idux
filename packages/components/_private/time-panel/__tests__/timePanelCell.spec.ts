import { MountingOptions, mount } from '@vue/test-utils'
import { computed } from 'vue'

import { renderWork } from '@tests'

import TimePanelCell from '../src/TimePanelCell'
import { timePanelContext } from '../src/tokens'
import { TimePanelCellProps } from '../src/types'

const mountGlobalOpts = {
  provide: {
    [timePanelContext as symbol]: {
      mergedPrefixCls: computed(() => 'ix-time-panel'),
    },
  },
}

describe('TimePanelCell', () => {
  const TimePanelCellMount = (options?: MountingOptions<Partial<TimePanelCellProps>>) =>
    mount(TimePanelCell, {
      ...(options as MountingOptions<TimePanelCellProps>),
      global: mountGlobalOpts,
    })

  renderWork<TimePanelCellProps>(TimePanelCell, {
    global: mountGlobalOpts,
    props: {
      value: 1,
      disabled: false,
      selected: true,
    },
  })

  test('onActive work', async () => {
    const onActive = vi.fn()
    const wrapper = TimePanelCellMount({ props: { value: 'cell', disabled: false, selected: false, onActive } })
    wrapper.trigger('click')

    expect(onActive).toBeCalledWith({ value: 'cell', disabled: false })
  })
  test('disabled', async () => {
    const onActive = vi.fn()
    const wrapper = TimePanelCellMount({ props: { value: 'cell', disabled: true, selected: false, onActive } })
    wrapper.trigger('click')

    expect(wrapper.element.classList.contains('ix-time-panel-cell-disabled')).toBeTruthy()
    expect(onActive).not.toBeCalled()
  })

  test('selected', async () => {
    const onActive = vi.fn()
    const wrapper = TimePanelCellMount({ props: { value: 'cell', disabled: false, selected: true, onActive } })
    wrapper.trigger('click')

    expect(wrapper.element.classList.contains('ix-time-panel-cell-selected')).toBeTruthy()
    expect(onActive).not.toBeCalled()
  })
})
