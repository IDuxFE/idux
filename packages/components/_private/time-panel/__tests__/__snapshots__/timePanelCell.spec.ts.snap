// Vitest Snapshot v1

exports[`TimePanelCell > render work 1`] = `"<li class=\\"ix-time-panel-cell ix-time-panel-cell-selected\\">01</li>"`;
